### PR TITLE
Add New Completion API support

### PR DESCRIPTION
**Pull request by KirillOsenkov**
This salvages our long-running editorCompletion branch and adds a CompletionUIProvider to hook up the new completion to our UI.

### CODE REVIEW

**Comment by KirillOsenkov** [comment]
@DavidKarlas if you have time, could you please help with landing this? We're going to need the new Async Completion API soon for XAML, so it's good to save what we have.

**Comment by KirillOsenkov** [comment]
For reference, I've also pushed these branches:
https://github.com/mono/monodevelop/tree/editorCompletion
https://github.com/mono/monodevelop/tree/editorCompletionSquashed
https://github.com/mono/monodevelop/tree/editorCompletionRebased

but I'm not sure if I did the conflict resolution right in any of them. This branch just manually reapplies the changes.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Extensions/StockIconCodon.cs [hunk 1]
```diff
@@ -80,5 +84,14 @@ internal class StockIconCodon : ExtensionNode
 		public string Animation {
 			get { return animation; }
 		}
+
+		public ImageId ImageId {
+			get {
+				if (imageid == null)
+					return default (ImageId);
+				var semicolonIndex = imageid.IndexOf (';');
+				return new ImageId (Guid.Parse (imageid.Remove (semicolonIndex)), int.Parse (imageid.Substring (semicolonIndex + 1)));
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Extensions/StockIconCodon.cs]: I think what you wanted here was:
`Guid.Parse (imageid.Substring (0, semicolonIndex))`

Comment (DavidKarlas) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Extensions/StockIconCodon.cs]: Why, what is difference?

Comment (Therzok) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Extensions/StockIconCodon.cs]: Ohhhh, this is `Remove(int)`. Got it, nevermind!

File: main/src/addins/MonoDevelop.SourceEditor2/MonoDevelop.SourceEditor.VsCompletion/DefaultCompletionService.cs [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+using Microsoft.VisualStudio.Text.Editor;
+using System.Threading;
+
+namespace Microsoft.VisualStudio.Language.Intellisense.Implementation
+{
+	[Export (typeof (IAsyncCompletionItemSourceProvider))]
+	[Name ("Debug completion item source")]
+	[Order (After = "default")]
+	[ContentType ("any")]
+	public class DebugCompletionItemSourceProvider : IAsyncCompletionItemSourceProvider
```
Comment (Therzok) [inline main/src/addins/MonoDevelop.SourceEditor2/MonoDevelop.SourceEditor.VsCompletion/DefaultCompletionService.cs]: Should these be public?

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Composition/EditorExperimentationServiceInternal.cs [hunk 3]
```diff
@@ -1,11 +1,19 @@
...
 	internal class EditorExperimentationServiceInternal : IExperimentationServiceInternal
 	{
-		public bool IsCachedFlightEnabled (string flightName) => false;
+		public bool IsCachedFlightEnabled (string flightName) => flightName == "CompletionAPI";
+	}
+
+	[ExportWorkspaceService (typeof (IExperimentationService), ServiceLayer.Host), Shared]
+	class ExperimentationService : IExperimentationService
+	{
+		public bool IsExperimentEnabled (string experimentName) => true;
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Composition/EditorExperimentationServiceInternal.cs]: Uhm, we already have this on the Host layer:
https://github.com/xamarin/md-addins/blob/e389a1a0b8950c30d7336005360e466e13e1054b/Xamarin.Ide/Xamarin.Ide/Xamarin.Ide.Insights/XSExperimentationService.cs#L19

File: main/src/addins/MonoDevelop.SourceEditor2/MonoDevelop.SourceEditor.VsCompletion/DefaultCompletionService.cs (hunk 2 above)
Comment (KirillOsenkov) [reply main/src/addins/MonoDevelop.SourceEditor2/MonoDevelop.SourceEditor.VsCompletion/DefaultCompletionService.cs]: Yeah probably not. Good catch.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Composition/EditorExperimentationServiceInternal.cs (hunk 3 above)
Comment (KirillOsenkov) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Composition/EditorExperimentationServiceInternal.cs]: Ah, nice, then no need to include this. Will fix.

**Comment by KirillOsenkov** [comment]
LOL as it is right now, Completion started showing up for C#, overlapping the current MonoDevelop completion :) We'd need a way to control this and only turn on for the languages that want it, for now.

**Comment by KirillOsenkov** [comment]
OK, I've turned the new completion off. This should be ready to go in.

**Comment by KirillOsenkov** [comment]
I've tested completion in C#, JS, TS, .json, .cshtml, .css and it all seems to be working fine. We can merge this.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Editor.Extension/VsEditorCommandingTextEditorExtension.cs [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+		{
+			base.Initialize ();
+			view = Editor.TextView;
+			editorCommandHandlerService = CompositionManager.GetExportedValue<IEditorCommandHandlerServiceFactory> ().GetService (view);
+		}
+
+		[CommandUpdateHandler (TextEditorCommands.ShowCompletionWindow)]
+		void UpdateCompletionCommand (CommandInfo info)
+		{
+			info.Enabled = editorCommandHandlerService.GetCommandState ((textView, textBuffer) => new CommitUniqueCompletionListItemCommandArgs (textView, textBuffer), null).IsAvailable;
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.Editor.Extension/VsEditorCommandingTextEditorExtension.cs]: Not a big fan of this whole duplication of code here, but works for now.

File: main/src/addins/MonoDevelop.SourceEditor2/MonoDevelop.SourceEditor.VsCompletion/CompletionUI.cs [hunk 5]
```diff
@@ -0,0 +1,677 @@
...
+					if (ypos >= height)
+						return false;
+					if (ypos < -rowHeight)
+						return true;
+					xpos = iconTextSpacing;
+					var selected = index == SelectedItemIndex;
+					bool drawIconAsSelected = SelectionEnabled && selected;
+					var item = filteredItems [index];
+					string markup = GLib.Markup.EscapeText (item.CompletionItem.DisplayText);
+					string description = "";//TODO: David DataProvider.GetDescription (item, drawIconAsSelected);
```
Comment (Therzok) [inline main/src/addins/MonoDevelop.SourceEditor2/MonoDevelop.SourceEditor.VsCompletion/CompletionUI.cs]: Is this needed? Looks like the description is requested async and shown in its own window?

**Comment by KirillOsenkov** [comment]
Maybe worth mentioning some quick background about this PR. This is by no means the final feature, but more of a starting point for @mgoertz-msft to investigate the XAML language service. The code here is not important and will most likely go away, what is important is a means of targeting the new Completion API in VS for Mac by addins. It's probably not a great time investment in cleaning this up right now, we just want to unblock some partner teams that want to start experimenting with the new API.

P.S. I'm on vacation next week, so hopefully this can land so that @mgoertz-msft and @decriptor can experiment.

**Comment by KirillOsenkov** [comment]
@slluis what do you think about this PR?

**Comment by slluis** [comment]
All this code is going to be replaced by the new editor, and the XAML integration would be for VSMac 8.0, so I don't see the need of merging this branch. If this branch is only for doing integration research, this can be done using builds from this branch.

**Comment by decriptor** [comment]
@KirillOsenkov if that is the case maybe we can keep this branch/PR going, but with the `VsEditorCommandingTextEditorExtension` enabled.

**Comment by KirillOsenkov** [comment]
@mgoertz-msft FYI

**Comment by KirillOsenkov** [comment]
This is now obsolete and no longer necessary. @DavidKarlas

**Comment by KirillOsenkov** [comment]
Keeping the branch just in case until we ship 8.3

**Pull request closed without merging**